File: frontend/src/config/sla.ts
export interface SLAConfig {
  hours: number;
  label: string;
}

// Configurable SLA thresholds across severity tiers
export const SLA_THRESHOLDS: Record<string, SLAConfig> = {
  CRITICAL: {
    hours: 24, // 24 hours
    label: '24h SLA',
  },
  HIGH: {
    hours: 72, // 3 days
    label: '3d SLA',
  },
  MEDIUM: {
    hours: 168, // 7 days
    label: '7d SLA',
  },
  LOW: {
    hours: 336, // 14 days
    label: '14d SLA',
  },
};

export type SLABreachStatus = 'OK' | 'DUE_SOON' | 'BREACHED';

export function getSLABreachStatus(severity?: string, openedAt?: string | Date | null): {
  status: SLABreachStatus;
  ageHours: number;
  slaHours: number;
  remainingHours: number;
  badgeLabel: string;
} {
  const sevKey = (severity || 'MEDIUM').toUpperCase();
  const sla = SLA_THRESHOLDS[sevKey] || SLA_THRESHOLDS.MEDIUM;

  if (!openedAt) {
    return {
      status: 'OK',
      ageHours: 0,
      slaHours: sla.hours,
      remainingHours: sla.hours,
      badgeLabel: `0h / ${sla.label}`,
    };
  }

  const opened = new Date(openedAt).getTime();
  const now = Date.now();
  const ageHours = Math.max(0, (now - opened) / (1000 * 60 * 60));
  const remainingHours = sla.hours - ageHours;

  let status: SLABreachStatus = 'OK';
  if (remainingHours <= 0) {
    status = 'BREACHED';
  } else if (remainingHours <= sla.hours * 0.25) {
    status = 'DUE_SOON';
  }

  let badgeLabel = `${Math.floor(ageHours)}h old (${sla.label})`;
  if (status === 'BREACHED') {
    badgeLabel = `BREACHED (+${Math.abs(Math.floor(remainingHours))}h)`;
  } else if (status === 'DUE_SOON') {
    badgeLabel = `DUE SOON (${Math.floor(remainingHours)}h left)`;
  }

  return {
    status,
    ageHours: Math.round(ageHours * 10) / 10,
    slaHours: sla.hours,
    remainingHours: Math.round(remainingHours * 10) / 10,
    badgeLabel,
  };
}

File: backend/config/sla.py
from datetime import datetime
from typing import Dict, Any, Optional

# Simple SLA thresholds configurable by administrators
SLA_THRESHOLDS_HOURS = {
    "CRITICAL": 24,   # 24 hours
    "HIGH": 72,       # 3 days
    "MEDIUM": 168,    # 7 days
    "LOW": 336        # 14 days
}

def get_sla_info(severity: Optional[str], opened_at: Optional[datetime], now: Optional[datetime] = None) -> Dict[str, Any]:
    if now is None:
        from backend.utils.time_utils import utc_now
        now = utc_now()
    
    sev_key = (severity or "MEDIUM").upper()
    sla_hours = SLA_THRESHOLDS_HOURS.get(sev_key, 168)
    
    if not opened_at:
        return {
            "status": "OK",
            "age_hours": 0.0,
            "sla_hours": sla_hours,
            "remaining_hours": float(sla_hours)
        }
        
    age_hours = round(max(0.0, (now - opened_at).total_seconds() / 3600.0), 1)
    remaining = round(sla_hours - age_hours, 1)
    
    if remaining <= 0:
        status = "BREACHED"
    elif remaining <= (sla_hours * 0.25):
        status = "DUE_SOON"
    else:
        status = "OK"
        
    return {
        "status": status,
        "age_hours": age_hours,
        "sla_hours": sla_hours,
        "remaining_hours": remaining
    }
